firewood/src/merkle.rs

```diff
@@ -473,385 +440,356 @@ impl<T: WriteLinearStore> Merkle<T> {
+                };
+
+                // `last_node_leaf` is at a strict prefix of `key`. Replace it with a branch.
+                //
+                //     ...                ...
+                //      |      -->         |
+                //  last_node       updated_last_node
+                //                         |
+                //                      new_leaf
+                let new_leaf_addr = self.create_node(Node::Leaf(LeafNode {
+                    value,
+                    partial_path: Path::from_nibbles_iterator(remaining_key),
+                }))?;
+
+                let mut updated_last_node: BranchNode = last_node_leaf.into();
+                updated_last_node.update_child(child_index, Some(new_leaf_addr));
+
+                self.update_node(
+                    traversal_path.iter().rev(),
+                    last_node.addr,
+                    Node::Branch(Box::new(updated_last_node)),
+                )?;
+
+                Ok(hash_invalidation_addresses)
+            }
+            Node::Branch(last_node_branch) => {
+                let Some(&last_node_to_child_index) = remaining_key.first() else {
+                    // `last_node_branch` is at `key`. Update its value.
+                    //     ...                ...
+                    //      |      -->         |
+                    //  last_node       updated_last_node
+                    let updated_last_node = Node::Branch(Box::new(BranchNode {
                         children: last_node_branch.children,
                         partial_path: last_node_branch.partial_path.clone(),
                         value: Some(value),
                         child_hashes: last_node_branch.child_hashes.clone(),
                     }));
 
-                    // TODO we can't use this macro because the node may move
-                    update_always_shrinks!(self.update_node(last_node.addr, new_branch))?;
-                }
-                Node::Leaf(last_node_leaf) => {
-                    // Replace the value in the leaf node.
-                    let new_leaf = Node::Leaf(LeafNode {
+                    self.update_node(
+                        traversal_path.iter().rev(),
+                        last_node.addr,
+                        updated_last_node,
+                    )?;
+
+                    return Ok(hash_invalidation_addresses);
+                };
+                // `last_node` is at a strict prefix of `key`.
+                // `remaining_key` is `key` after removing the prefix of `last_node` and 1 additional
+                // nibble for the child index that points to the node after `last_node`.
+                let remaining_key = &remaining_key[1..];
+
+                // See if the `last_node` has a child where `key` would go.
+                let child_option = *last_node_branch
+                    .children
+                    .get(last_node_to_child_index as usize)
```

Comment:
`child_option` is not an option.

nit: we should think about better helper methods in `BranchNode` so we can just do: `last_node_branch.child(last_node_to_child_index)`. If we do this, then the `mut_child` should be `child_mut`.

Reply:
Added `child` method and renamed `mut_child` to `child_mut`